### PR TITLE
build: Scripts to manage release process

### PR DESCRIPTION
**Pull request by sandrina-p**
Added the scripts that we must run to publish a new release.

- `npm run release:dev:patch` -> To publish dev releases in PRs
- `npm run release:main:patch` -> To publish stable releases in main.

Updated the `CONTRIBUTING.md` explaining this development flow.

- Loom: [Development overview](https://www.loom.com/share/34554039380648b1b2922333d4e4b51a) - basically I do the steps written in the CONTRIBUTING file. 
- Loom: [scripts walkthrough](https://www.loom.com/share/4955fe2845d542dda327d8d9565e64a0) - a detailed overview of the scripts created.

Important:
- **DO NOT** test it in this repo, otherwise you'll publish unwanted versions. Use https://github.com/remoteoss/test-pkg-npm instead.
- Before you test, please copy the `scripts/release*` from here to there, to ensure you have the latest updates.

### CODE REVIEW

File: CONTRIBUTING.md
Comment (brennj) [inline CONTRIBUTING.md:9]: Glad we're being explicit about this

File: scripts/release_dev.js
Comment (brennj) [inline scripts/release_dev.js:24]: <3

File: scripts/release_main.js
Comment (brennj) [inline scripts/release_main.js:105]: dumb q. how do you protect `main` when you're doing this?

Comment (sandrina-p) [reply scripts/release_main.js:105]: We don't. It's shitty. Ideally, this should happen after a feature PR is merged:
- GH Action triggers, which opens a new MR with the new version + CHANGELOG.
- You (or any maintainer) review the MR and approve it. Or tweak it if needed, or even close it, in case it doesn't make sense to release a new version.
  - On merge, another action is triggered to do the actual release.

But as I don't know Github Actions I couldn't make it happen :(

File: scripts/release_dev.js
Comment (brennj) [inline scripts/release_dev.js:19]: do you still need the await if its `Sync`?

**Review by sandrina-p — COMMENTED** [review]
@johnstonbl01 and @johnstonbl01 this is (finally) ready for review!

P.S. This is pointing to `copy-sourcecode` branch so it benefits from the eslint friends. I'll rebase it in a few minutes after that PR is merged.

File: scripts/release.helpers.js
Comment (brennj) [inline scripts/release.helpers.js:12]: lots of good defensive programming here that we need

File: scripts/release_dev.js
Comment (sandrina-p) [reply scripts/release_dev.js]: No, this was old code, it's gone

File: scripts/release.helpers.js
Comment (sandrina-p) [reply scripts/release.helpers.js:12]: This PR is the biggest proof I'm human. Each `if` was a mistake I've done. 😅

File: scripts/release_dev.js
Comment (sandrina-p) [reply scripts/release_dev.js:1]: Good point. Yes, because the branch exists in the origin. So it's still public/visible even though it doesn't has an explicit PR.

File: scripts/release_main.js
Comment (brennj) [inline scripts/release_main.js:49]: `beta.0` is maybe a little weird to me but I'm trying to get my thoughts correct here:

- We want to keep it unstable until we're ready for a correct `1.0.0`.
- So then wouldn't we be bumping the `beta.X` instead? and its `1.0.0.beta.123` until we're happy? 
- Thats not necessarily semver though... but for all patches, it should maybe do that instead?

Comment (sandrina-p) [reply scripts/release_main.js:49]: Honestly I'm not sure either, so here's my answer:

`beta.0` -> it's mandatory to have a number after any prefix. it cannot be just `.beta`

- My idea would be to release `0.1.0-beta.0` and then `0.x.x-beta.0` as we iterate. I'd prefer that over `.beta-x` so it's clear when new releases are features or fixes. Those this make sense?
- For that reason, I don't see us bumping `beta.x`. AFAIK `beta.x` is used to fix bugs directly related to that given version, not to fix bugs or release new feature in general.

Let me try to give an example. We start with `0.1.0-beta.0`.
- `0.2.0-beta.0` -> we now support this "foo1" thing.
- `0.2.0-beta.0` -> we now support this "bar" thing.
- `0.2.1-beta.0` -> we now fixed a bug at "biz" thing.

See, each commit it's about different features within the library as a whole. If we used `0.1.0-beta.x`, then it would be like:
- `0.2.0-beta.0` -> we now support this "foo1" thing.
- `0.2.0-beta.1` -> we now support this "foo2" thing.
- `0.2.0-beta.2` -> we now fixed a bug at "bar" thing.

It's not clear if the new versions are fixes, features, etc. 

In my understanding `beta` is used in the following scenario:
- `0.2.0-beta.0` -> we now support this "foo" thing.
- `0.2.0-beta.1` -> we fixed a bug related to the "foo" thing support this "foo2" thing.
- `0.2.0-beta.2` -> we fixed another bug related to the "foo" thing.
- `0.2.0-beta.3` -> we need to ship "bar" this. Does it make sense to be here? it should be "0.3.0-beta.0" instead.

Does this make sense? Honestly I don't know, I didn't find any document super clear about this, and looking at the NPM ecosystem around didn't bring me clarity either :/

Comment (brennj) [reply scripts/release_main.js:49]: > beta.0 -> it's mandatory to have a number after any prefix. it cannot be just .beta

TIL 💡 

My idea would be to release 0.1.0-beta.0 and then 0.x.x-beta.0 as we iterate. I'd prefer that over .beta-x so it's clear when new releases are features or fixes. Those this make sense?
For that reason, I don't see us bumping beta.x. AFAIK beta.x is used to fix bugs directly related to that given version, not to fix bugs or release new feature in general.

Thats perfectly fine then, lets roll with it!

Comment (sandrina-p) [reply scripts/release_main.js:49]: Cool! Let me just rewrite `main` to add the co-authors, then rebase #2, and finally come back to this one. Give me ~10min.

File: CONTRIBUTING.md
Comment (brennj) [inline CONTRIBUTING.md:70]: Confirm for me:

It does not matter if a dev version gets into `main`, as long as it's followed up with the actual stable release. I don't need to necessarily revert any commits with the `package-lock.json`/`package.json` changes back to what the current stable version is.

**Comment by brennj** [comment]
i'm happy with the steps as our first iteration, there is a lot of manual work - but this is ok for now. I will recheck later when its rebased. @johnstonbl01 defo will appreciate your eyes here too

File: CONTRIBUTING.md
Comment (brennj) [reply CONTRIBUTING.md:70]: perfect 4 me

File: scripts/release_dev.js
Comment (johnstonbl01) [inline scripts/release_dev.js:33]: I think this is because this command is returning the _next to last_ hash. 

I think `git describe --always` or `git log -1 --abbrev-commit --pretty=oneline` should get you the correct hash?

**Comment by johnstonbl01** [comment]
I think there's a lot of opportunities here to automate some of this away, but for now having the manual processes is fine. 👍 Looks good to me, Sandrina!

File: scripts/release_dev.js
Comment (sandrina-p) [reply scripts/release_dev.js:33]: Hum no. `git describe --always` returns the hash of the last commit. Here I want to get the hash of the next (upcoming not yet done) commit. :/

Comment (sandrina-p) [reply scripts/release_dev.js:33]: Now that I have the approval of both John and Blake, shall we try to create `@0.1.0-dev.xxxx` in this PR or in another PR? 😅 I'd say let's go with this one.

Comment (johnstonbl01) [reply scripts/release_dev.js:33]: > Hum no. `git describe --always` returns the hash of the last commit. Here I want to get the hash of the next (upcoming not yet done) commit. :/

Ah ok. It's not a blocker, but think we should land on something we want to keep for the long-term soon-ish.

Comment (sandrina-p) [reply scripts/release_dev.js:33]: The alternative, which was my first commit, was to use date instead. eg `0.1.0-dev.20230517172340` (yyyymmddhhmmss). I've seen this in a couple of other libraries in the wild (eg [this one](https://www.npmjs.com/package/@pandacss/node?activeTab=versions)). WDYT?

Comment (johnstonbl01) [reply scripts/release_dev.js:33]: 🤔 I kind of like the date + timestamp tbh.

Comment (sandrina-p) [reply scripts/release_dev.js]: I agree. Sorry @brennj I'm turning back to the date again. https://github.com/remoteoss/json-schema-form/pull/3/commits/efe38a131727ea7cd7e2141ae52187d98d5d85a5 Yesterday you asked for the advantages, here's the answer:

- git sha do not match with the upcoming commit. On top of that, they are not historically sortable.
- By using the yyyymmddhhmmss format, we know their order of releases.

Comment (brennj) [reply scripts/release_dev.js]: i'm ok with that ;)

**Review by sandrina-p — COMMENTED** [review]
Check my latest commit. https://github.com/remoteoss/json-schema-form/pull/3/commits/28792ddde1d9b12a50c36829f96e9996c941f079 I noticed a "missing feature" where the actual package would be published empty. Why? Because it was missing running the `build` script to generate the `dist/` during the release script.

File: scripts/release_dev.js
Comment (sandrina-p) [inline scripts/release_dev.js:105]: note: Check my latest commit. https://github.com/remoteoss/json-schema-form/pull/3/commits/28792ddde1d9b12a50c36829f96e9996c941f079 I noticed a "missing feature" where the actual package would be published empty. Why? Because it was missing running the `build` script to generate the `dist/` during the release script.

Comment (sandrina-p) [reply scripts/release_dev.js:105]: How you can test this (in this repo safely):

- run `npm run release:local` - and unzip the generated tarball. (this is explained in the read the CONTRIBUTING too)

To test the `npm run release:dev:*` or `npm run release:main`

- Comment the line with `await checkGitBranchAndStatus()` to bypass git checks
- Comment the line after the `await build` to ensure it doesn't try to commit or publish.

Comment (brennj) [reply scripts/release_dev.js:105]: on it now

Comment (sandrina-p) [reply scripts/release_dev.js:105]: Forgot to say, but I also created the `relese:local` script right before that commit, here: https://github.com/remoteoss/json-schema-form/pull/3/commits/fbadf2ea3039c3fcc0075dad1483a79cd08d360b

Comment (brennj) [reply scripts/release_dev.js:105]: - `npm run release:local` works lovely and i've tested it in dragon and it shows up all in its glory.
- with the scary code behind comments, the `release:*:*` also are there... 

TBH for me this is good, 🚢 ?

**Review by brennj — APPROVED** [review]

File: scripts/release_dev.js
Comment (sandrina-p) [reply scripts/release_dev.js:105]: Outcome from our Slack huddle:
- The first [dev release](https://www.npmjs.com/package/@remoteoss/json-schema-form) was a success!
- We open an MR in our internal codebase to replace the local imports with the released package.
- We noticed a couple of missing exports : (https://github.com/remoteoss/json-schema-form/pull/3/commits/8116f6fb7947e7c31d770681a5e5c488f8561bd5 and https://github.com/remoteoss/json-schema-form/pull/3/commits/8775cbf454a161de41ebd25b0f237a36e23af76f)

Tomorrow we will:
- Finish replacing the imports
- Wait for a green E2E in our internal codebase
- Merge this branch
- Release beta version in `main` :excited:

File: scripts/release.helpers.js
Comment (tugayac) [inline scripts/release.helpers.js:59]: Need to pass in `question` here again.

File: scripts/release.helpers.js
Comment (tugayac) [inline scripts/release.helpers.js:45]: The uppercase "Y" here implies the default is "yes" (so if I press enter, I expect to confirm, without having to type "Y"). However, I get an error:
```
Invalid input. Please enter Y or n.
```

If you want to prevent accidental entries, we can change it to just `y/n`?

File: scripts/release_dev.js
Comment (tugayac) [inline scripts/release_dev.js:50]: In case you were wondering why the commit SHA in the dev release version wasn't matching the current commit at the time the script is executed, this is why: https://www.loom.com/share/d5a02b42b04b407195e2b30311ba32c1

File: scripts/pr_dev_version.js
Comment (tugayac) [inline scripts/pr_dev_version.js:16]: Depending on what you want to check in terms of diffs, you might be able to use `git diff` with some flags to achieve what you want.

For example, you can do:
```
git diff --name-status --merge-base main HEAD
```
which will show what happened to the file + the file name, something like this:
```
M       mix.exs
D       priv/json_schemas/common/contract_details/order_all.json
M       priv/json_schemas/contract_details/AUS.json
A       priv/json_schemas/contract_details/AUT.json
```
It essentially shows what changed between the current HEAD and the commit that's the common ancestor of the current branch and `main` (assuming there's only a single common ancestor, which should be the case for us)

File: scripts/release_dev.js
Comment (sandrina-p) [reply scripts/release_dev.js:50]: Nice TIL but you missed the point I want the SHA of the next commit (not existing yet). For example:

```bash
- 19ea0c1 "Release v0.1.0-dev.19ea0c1"
- 575f1a0 "feat: add potatoes"
- 9e21c3 "feat: add potatoes"
````

I need to know "19ea0c1" before I commit it so it matches in the commit msg.

But either way, even  if we find the answer, i'd rather stick with the date hash (yyyy..ss) because it's chronological and it was already useful yesterday to verify some versions.

File: scripts/pr_dev_version.js
Comment (sandrina-p) [reply scripts/pr_dev_version.js]: Today with fresh eyes I found a (different) way! Instead of using `git diff` to get the difference, I used `git show` to output the file content of main, and then parse it and manually compare it to the current branch. Check it out at https://github.com/remoteoss/json-schema-form/pull/3/commits/20b69f27eb14dcba86698766315a9538f9c8ae16

File: scripts/release.helpers.js
Comment (sandrina-p) [reply scripts/release.helpers.js]: Good point, fixed at https://github.com/remoteoss/json-schema-form/pull/3/commits/6019938c955c16636ea02e15a01640572db02563

File: scripts/release.helpers.js
Comment (sandrina-p) [reply scripts/release.helpers.js]: Excellent catch, fixed https://github.com/remoteoss/json-schema-form/pull/3/commits/6019938c955c16636ea02e15a01640572db02563

File: scripts/pr_dev_version.js
Comment (sandrina-p) [reply scripts/pr_dev_version.js]: Ah, although this worked locally, it [fails in CI](https://github.com/remoteoss/json-schema-form/actions/runs/5012270770/jobs/8984036926?pr=3) because it can't perform `git` commands. So I reverted the `pr_dev_version.js` to what I had before https://github.com/remoteoss/json-schema-form/pull/3/commits/241dfc2a37c966cbb7b5022f65b08ed2b4820716.

Comment (sandrina-p) [reply scripts/pr_dev_version.js]: Okay, now is working file, with [CI throwing the expected error](https://github.com/remoteoss/json-schema-form/actions/runs/5012509498/jobs/8984543336?pr=3). I've pushed a last commit to revert it back to `0.0.0` using `npm run version_as_main`. https://github.com/remoteoss/json-schema-form/pull/3/commits/241dfc2a37c966cbb7b5022f65b08ed2b4820716

File: scripts/release_dev.js
Comment (tugayac) [reply scripts/release_dev.js:50]: Ah got it! Yeah I like the date-based version, that's a good idea 💡

**Pull request merged**
